### PR TITLE
Implement heatmap tracks in oncoprinter, and export heatmaps to oncoprinter from results page

### PR DESCRIPTION
**Pull request by adamabeshouse**

### CODE REVIEW

**Comment by jjgao** [comment]
@adamabeshouse Looks good. Thanks!

For the format text, can we still do it in the old way, ie. expand and collapse?

**Comment by adamabeshouse** [comment]
@jjgao we had a whole long convo about that the other day and decided on this 😅  is it okay to leave it like this and change it if theres complaints?

**Comment by jjgao** [comment]
@adamabeshouse Do you remember the reasoning that it was decided on this? 

I don't feel strongly but would like to listen to input from others.

**Comment by adamabeshouse** [comment]
@jjgao I think the main idea was that it may be easier to use if you can open the instructions in a new window. But I don't remember very well. @inodb @alisman do you remember?

**Comment by adamabeshouse** [comment]
@tmazor how does it look to you?

**Comment by tmazor** [comment]
The new layout looks great!

I think I have a slight preference for the old approach of expand/collapse for the format descriptions rather than a separate page, but I don't feel super strongly either way. Totally ok with me to leave as is with the new page.

Just one comment about the heatmaps - can the legend name be the same as the track name? When I add the example data, the legend says "Heatmap" 3 times but it would be clearer if they had the same name as the data tracks:
image

**Comment by adamabeshouse** [comment]
Great, thanks @tmazor ! the only concern with that is probably often people will have more than one heatmap track on the same color scale. In the results view, the legend will have the name of the alteration type (e.g. Expression Heatmap), as opposed to the names of the tracks (e.g. KRAS / PTEN / BRAF). I guess we could make it different for the special case of one heatmap track but... idk. What do you think?

**Comment by tmazor** [comment]
Ohhh I see, that makes sense. Maybe it's ok to leave as is - the 3 color schemes are different enough that's it's fairly obvious which is which, and this is user-provided data so they should be at least a little familiar with the data they provided. 

Or, what about naming the legend similar to the data format? e.g. Heatmap (z-scores), Heatmap (0-1), Heatmap

**Comment by adamabeshouse** [comment]
I think naming by color scale type, as you suggested, would work well!

**Comment by adamabeshouse** [comment]
@tmazor okay, I made all those changes!

**Comment by tmazor** [comment]
looks good to me @adamabeshouse !

File: src/pages/staticPages/tools/oncoprinter/styles.module.scss [hunk 1]
```diff
@@ -1,17 +1,16 @@
...
+    max-width: 800px;
 
     .dataFormatHelpBorder {
         border-left: 1px solid black;
         margin-left: 15px;
         margin-right: 15px;
     }
 }
+
+.inputSection {
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/styles.module.scss:15]: ?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts [hunk 2]
```diff
@@ -0,0 +1,572 @@
...
+
+type OncoprinterHeatmapTrackSpec = {
+    trackName: string;
+    datatype: HeatmapTrackDataType;
+};
+
+type OncoprinterClinicalTrackDatum = Pick<
+    ClinicalTrackDatum,
+    'attr_id' | 'uid' | 'attr_val_counts' | 'na' | 'attr_val'
+> & { sample: string };
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts:36]: sampleId?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+type OncoprinterHeatmapTrackDatum = Pick<
+    IBaseHeatmapTrackDatum,
+    'profile_data' | 'uid' | 'na'
+> & { sample: string };
+
+// Output mimics mobxpromise form but thats just because its a nice way to package status and result.
+// This isn't meant to be plugged into mobxpromise machinery i.e. with `await`
+type ParseInputResult<TrackSpecType> =
+    | {
+          status: 'complete';
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts:47]: 'complete' implies to me some kind of async process and could lead to confusion given mobxpromise.    could we use 'parsedSucess' or 'success' or 'valid' ?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterConstants.ts [hunk 4]
```diff
@@ -592,373 +597,406 @@ TCGA-25-2392-01
...
-TCGA-25-1877-01\t43\t${ONCOPRINTER_CLINICAL_VAL_NA}\t24\t${ONCOPRINTER_CLINICAL_VAL_NA}
-TCGA-25-1878-01\t33\tProstate\t104\t${ONCOPRINTER_CLINICAL_VAL_NA}
-TCGA-25-2042-01\t50\tProstate\t49\t${ONCOPRINTER_CLINICAL_VAL_NA}
-TCGA-25-2390-01\t21\tProstate\t17\t${ONCOPRINTER_CLINICAL_VAL_NA}
-TCGA-25-2396-01\t51\tBreast\t1\t${ONCOPRINTER_CLINICAL_VAL_NA}
-TCGA-25-2397-01\t33\tBreast\t40\t${ONCOPRINTER_CLINICAL_VAL_NA}
-`;
+// generate using seeded random number so its the same every time but we don't have to type it out manually
+const rng = new DeterministicRNG();
+const samples = [
```
Comment (alisman) [inline src/pages/staticPages/tools/oncoprinter/OncoprinterConstants.ts:602]: exampleSampleList?

File: src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts (hunk 2 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts:36]: I'm just going along with the existing oncoprint type name which is `sample`, this just makes `sample` mandatory whereas in the oncoprint data type its optional

File: src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts (hunk 3 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterClinicalAndHeatmapUtils.ts:47]: good point, I will change that

File: src/pages/staticPages/tools/oncoprinter/OncoprinterConstants.ts (hunk 4 above)
Comment (adamabeshouse) [reply src/pages/staticPages/tools/oncoprinter/OncoprinterConstants.ts:602]: sure

**Pull request merged**